File: 0007/P0007.cpp
#include <cmath>
#include <iostream>
#include <vector>

bool isPrime(int n, std::vector<int>& primes) {
  int s = sqrt(n);
  
  for (int prime : primes) {
    if (n % prime == 0) return false;
    if (prime > s) return true;
  }
  
  return true;
}

int prime(int n, std::vector<int>& primes) {
  int nextPrime = primes[primes.size() - 1] + 2;
  while (primes.size() <= n) {
    if (isPrime(nextPrime, primes)) {
      primes.push_back(nextPrime);
    }
    nextPrime += 2;
  }

  return primes[n - 1];
}

int main() {
  std::vector<int> primes{2, 3};
  int t;
  std::cin >> t;
  for (int i = 0; i < t; ++i) {
    int n;
    std::cin >> n;
    int p = prime(n, primes);
    std::cout << p << std::endl;
  }

}

File: 0007/P0007.hs
{-# LANGUAGE UnicodeSyntax #-}

readInt ∷ IO Int
readInt = readLn

main ∷ IO ()
main = do
  t <- readInt
  countNthPrime t [3, 2]

countNthPrime ∷ Int → [Int] → IO ()
countNthPrime 0 _      = return ()
countNthPrime i primes = do
  n <- readInt
  let (p, primes') = prime n (head primes + 2) primes
  print p
  countNthPrime (i - 1) primes'

prime ∷ Int → Int -> [Int] → (Int, [Int])
prime n nextPrime primes
  | length primes >= n       = (primes!!(length primes - n), primes)
  | isPrime nextPrime primes = prime n (nextPrime + 2) (nextPrime : primes)
  | otherwise                = prime n (nextPrime + 2) primes

isPrime ∷ Int → [Int] → Bool 
isPrime n primes = isPrime' n (s n) (reverse primes)
  where
    s = floor . sqrt . fromIntegral
    isPrime' n s primes
      | null primes              = True
      | head primes > s          = True
      | n `mod` head primes == 0 = False
      | otherwise                = isPrime' n s (tail primes)
